Error compiling on Windows (in CI)

Windows CI job via AppVeyor fails with:

```
Build started
git clone -q https://github.com/havoc-io/go-keytar.git C:\projects\src\github.com\havoc-io\go-keytar
git fetch -q origin +refs/pull/2/merge:
git checkout -qf FETCH_HEAD
set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
cd %GOPATH%\src\github.com\havoc-io\go-keytar
go build
set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
cd %GOPATH%\src\github.com\havoc-io\go-keytar
go test
--- FAIL: TestNonExistentReplace (0.00s)
panic: runtime error: cgo argument has Go pointer to Go pointer [recovered]
	panic: runtime error: cgo argument has Go pointer to Go pointer
goroutine 21 [running]:
testing.tRunner.func1(0xc0000a8400)
	C:/go/src/testing/testing.go:830 +0x399
panic(0x540be0, 0xc00004c540)
	C:/go/src/runtime/panic.go:522 +0x1c3
github.com/havoc-io/go-keytar.(*keychainWindows).AddPassword.func2(0xc0000b8000, 0xc0000b8000)
	C:/projects/src/github.com/havoc-io/go-keytar/keytar_windows.go:66 +0x96
github.com/havoc-io/go-keytar.(*keychainWindows).AddPassword(0x671658, 0x56cae7, 0x17, 0x56cad0, 0x17, 0x56a618, 0x10, 0x0, 0x0)
	C:/projects/src/github.com/havoc-io/go-keytar/keytar_windows.go:66 +0x28d
github.com/havoc-io/go-keytar.ReplacePassword(0x590d20, 0x671658, 0x56cae7, 0x17, 0x56cad0, 0x17, 0x56a618, 0x10, 0x4c4764, 0x62bac8)
	C:/projects/src/github.com/havoc-io/go-keytar/keytar.go:66 +0xc0
github.com/havoc-io/go-keytar.TestNonExistentReplace(0xc0000a8400)
	C:/projects/src/github.com/havoc-io/go-keytar/keytar_test.go:54 +0xb0
testing.tRunner(0xc0000a8400, 0x572e08)
	C:/go/src/testing/testing.go:865 +0xc7
created by testing.(*T).Run
	C:/go/src/testing/testing.go:916 +0x361
exit status 2
FAIL	github.com/havoc-io/go-keytar	0.028s
Command exited with code 1
```

--- a/README.md
+++ b/README.md
@@ -122,3 +122,8 @@ was created by another program not enforcing this constraint.
   codes.
 - Figure out if Go has a secure fallback that we could use somewhere in its
   crypto libraries.
+  
+## Contributors
+
+- Jacob Howard (@havoc-io)
+- Jeffrey Hulten (@jhulten)

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -24,7 +24,6 @@ clone_folder: C:\projects\go-keytar
 build_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
   - cd %PROJECTPATH%
-  - go get -t
   - go build
 
 # Run tests

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -23,6 +23,7 @@ clone_folder: C:\projects\src\github.com\havoc-io\go-keytar
 build_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
   - cd %GOPATH%\src\github.com\havoc-io\go-keytar
+  - go get -t
   - go build
 
 # Run tests

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -6,7 +6,7 @@ version: 1.0.0.{build}
 # http://help.appveyor.com/discussions/suggestions/615-support-for-msys2#comment_37571757
 environment:
   GOPATH: C:\projects\go
-  PROJECT_PATH: C:\projects\go-keytar
+  PROJECTPATH: C:\projects\go-keytar
 
   matrix:
   - GOROOT: C:\go-x86
@@ -17,20 +17,20 @@ environment:
     MINGWROOT: C:\msys64\mingw64
 
 # Set up clone path
-clone_folder: C:\projects\src\github.com\havoc-io\go-keytar
+clone_folder: C:\projects\go-keytar
 
 # Run build (primarily just here to stop the default Visual Studio build, since
 # go test will build anyway)
 build_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
-  - cd %PROJECT_PATH%
+  - cd %PROJECTPATH%
   - go get -t
   - go build
 
 # Run tests
 test_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
-  - cd %PROJECT_PATH%
+  - cd %PROJECTPATH%
   - go test
 
 # Send notifications

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,7 +5,8 @@ version: 1.0.0.{build}
 # NOTE: Structure of MSYS2 installation outlined in this comment:
 # http://help.appveyor.com/discussions/suggestions/615-support-for-msys2#comment_37571757
 environment:
-  GOPATH: C:\projects
+  GOPATH: C:\projects\go
+  PROJECT_PATH: C:\projects\go-keytar
 
   matrix:
   - GOROOT: C:\go-x86
@@ -22,14 +23,14 @@ clone_folder: C:\projects\src\github.com\havoc-io\go-keytar
 # go test will build anyway)
 build_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
-  - cd %GOPATH%\src\github.com\havoc-io\go-keytar
+  - cd %PROJECT_PATH%
   - go get -t
   - go build
 
 # Run tests
 test_script:
   - set PATH=%GOROOT%\bin;%MINGWROOT%\bin;%PATH%
-  - cd %GOPATH%\src\github.com\havoc-io\go-keytar
+  - cd %PROJECT_PATH%
   - go test
 
 # Send notifications

--- a/go.mod
+++ b/go.mod
@@ -1,0 +1,3 @@
+module github.com/havoc-io/go-keytar
+
+go 1.12

--- a/go.mod
+++ b/go.mod
@@ -1,3 +1,7 @@
 module github.com/havoc-io/go-keytar
 
 go 1.12
+
+require (
+    github.com/danieljoos/wincred v1.0.2
+)

--- a/keytar_windows.go
+++ b/keytar_windows.go
@@ -30,9 +30,7 @@ func (*keychainWindows) AddPassword(service, account, password string) error {
 
 	cred := wincred.NewGenericCredential(targetFormat(service, account))
 	cred.CredentialBlob = []byte(password)
-	err := cred.Write()
-
-	return err
+	return cred.Write()
 }
 
 func (*keychainWindows) GetPassword(service, account string) (string, error) {
@@ -64,12 +62,11 @@ func (*keychainWindows) DeletePassword(service, account string) error {
 	if err != nil {
 		return err
 	}
-	cred.Delete()
 
-	return nil
+	return cred.Delete()
 }
 
 func init() {
-	// Register the OS X keychain implementation
+	// Register the Windows keychain implementation
 	keychain = &keychainWindows{}
 }

--- a/keytar_windows.go
+++ b/keytar_windows.go
@@ -1,15 +1,10 @@
 package keytar
 
-// #include <stdlib.h>
-// #include <windows.h>
-// #include <wincred.h>
-import "C"
 
 import (
-	// System imports
 	"fmt"
-	"unsafe"
-	"syscall"
+
+	"github.com/danieljoos/wincred"
 )
 
 // Utility function to format service/account into something Windows can store
@@ -33,42 +28,11 @@ func (*keychainWindows) AddPassword(service, account, password string) error {
 		return ErrInvalidValue
 	}
 
-	// Compute target item name
-	target := targetFormat(service, account)
+	cred := wincred.NewGenericCredential(targetFormat(service, account))
+	cred.CredentialBlob = []byte(password)
+	err := cred.Write()
 
-	// Convert the target name.  We require that inputs be in UTF-8, but even
-	// then we can't use these using the Windows ANSI (A) APIs, so we have to
-	// use the Unicode (W) APIs, but these all use UTF-16, we we need to
-	// generate UTF-16 views of our strings.  Fortunately, the Windows syscall
-	// package has a nice API for doing this.
-	targetUTF16Ptr, err := syscall.UTF16PtrFromString(target)
-	if err != nil {
-		return ErrInvalidValue
-	}
-	targetUTF16 := C.LPWSTR(unsafe.Pointer(targetUTF16Ptr))
-
-	// Convert the password blob.  This is just stored as a raw array of bytes,
-	// so we can store it UTF-8 encoded.
-	passwordBlobSize := C.DWORD(len(password))
-	passwordVoidBlob := unsafe.Pointer(C.CString(password))
-	defer C.free(passwordVoidBlob)
-	passwordBlob := C.LPBYTE(passwordVoidBlob)
-
-	// Set up the credential
-	var credential C.CREDENTIALW
-	credential.Type = C.CRED_TYPE_GENERIC
-	credential.TargetName = targetUTF16
-	credential.CredentialBlobSize = passwordBlobSize
-	credential.CredentialBlob = passwordBlob
-	credential.Persist = C.CRED_PERSIST_LOCAL_MACHINE
-
-	// Store the credential
-	if C.CredWriteW(&credential, 0) != C.TRUE {
-		return ErrUnknown
-	}
-
-	// All done
-	return nil
+	return err
 }
 
 func (*keychainWindows) GetPassword(service, account string) (string, error) {
@@ -79,35 +43,13 @@ func (*keychainWindows) GetPassword(service, account string) (string, error) {
 		return "", ErrInvalidValue
 	}
 
-	// Compute target item name
-	target := targetFormat(service, account)
+	cred, err := wincred.GetGenericCredential(targetFormat(service, account))
 
-	// Convert the target name.  See note in AddPassword.
-	targetUTF16Ptr, err := syscall.UTF16PtrFromString(target)
 	if err != nil {
-		return "", ErrInvalidValue
-	}
-	// NOTE: For some reason they use LPCWSTR here, as opposed to LPWSTR in the
-	// CREDENTIALW struct
-	targetUTF16 := C.LPCWSTR(unsafe.Pointer(targetUTF16Ptr))
-
-	// Query the credential
-	var credential C.PCREDENTIALW
-	if C.CredReadW(targetUTF16, C.CRED_TYPE_GENERIC, 0, &credential) != C.TRUE {
-		return "", ErrNotFound
+		return "", err
 	}
 
-	// Extract the password blob
-	result := C.GoStringN(
-		(*C.char)(unsafe.Pointer(credential.CredentialBlob)),
-		C.int(credential.CredentialBlobSize),
-	)
-
-	// Free the credential memory
-	C.CredFree(C.PVOID(credential))
-
-	// All done
-	return result, nil
+	return string(cred.CredentialBlob), nil
 }
 
 func (*keychainWindows) DeletePassword(service, account string) error {
@@ -118,24 +60,12 @@ func (*keychainWindows) DeletePassword(service, account string) error {
 		return ErrInvalidValue
 	}
 
-	// Compute target item name
-	target := targetFormat(service, account)
-
-	// Convert the target name.  See note in AddPassword.
-	targetUTF16Ptr, err := syscall.UTF16PtrFromString(target)
+	cred, err := wincred.GetGenericCredential(targetFormat(service, account))
 	if err != nil {
-		return ErrInvalidValue
+		return err
 	}
-	// NOTE: For some reason they use LPCWSTR here, as opposed to LPWSTR in the
-	// CREDENTIALW struct
-	targetUTF16 := C.LPCWSTR(unsafe.Pointer(targetUTF16Ptr))
+	cred.Delete()
 
-	// Delete the credential
-	if C.CredDeleteW(targetUTF16, C.CRED_TYPE_GENERIC, 0) != C.TRUE {
-		return ErrUnknown
-	}
-
-	// All done
 	return nil
 }
 

